Expand motivation and add examples of subsuming other deriving extensions

Lysxia opened the pull request:
This addresses one of SPJ's concerns in the discussion thread.

TODO (unless you are planning do some of it already)

- [x] Improve the rigor of the change specification. (I copied part of it into the introduction so the motivation can be understood, but it was also requested that the actual specification be much more precise.)
- [x] Harmonize pronouns (we vs I) (In the end I'm keeping things as they are, on the grounds that the proposal will naturally take a less personal tone as it gets further refined.)
- [x] Merge with other existing examples. (I kinda just typed this out, but the last few paragraphs about `Traverse` overlap with what you already have.)
- [x] Mention "generic generic deriving". (We can make the dictionaries themselves instances of `Generic` and do stuff with that.)
- [x] Mention associated type families (a priori, as future work)

Lysxia committed:
Expand motivation and add examples of subsuming other deriving extensions

Lysxia commented:
If you're interested in merging this I can complete this patch.

tysonzero commented:
Definitely interested in merging. Awesome work, thank you!

Lysxia commented:
Ok I think that's it for this iteration. I'm sure there are going to be a few more. I tried to avoid actually changing the technical details; I don't agree with all of them, but we can discuss in the main thread once this is merged. The most significant addition after the motivation was to explain the new syntax for instances.

Lysxia commented:
~~Two~~ One additional points of discussion is ~~are the `deriving` syntax (needs specification), and~~ whether to specify the new `instance` declarations in terms of elaboration to Core (that actually seems better now that I think about it) instead of the existing surface syntax. I'm just writing that down here to remember to bring them up again in the main thread.

(EDIT: deriving syntax explained, see my next message below)

tysonzero commented:
It looks like the second half of the "Record types as classes" section can be removed, and any relevant parts of it can be added into the following "Instance declarations" section.

It may also be worth mentioning `DeriveAnyClass` (looks like I forgot to in the original).

Besides that it seems good to merge!

Lysxia commented:
Oh, I did not notice that part was made mostly redundant.

One extra thing it mentioned was the `deriving` syntax. I added a new section on that. There was a problem but I think I solved it while trying to explain it... Unless I'm missing something, the problem is that we must know the contexts of all instances before starting type inference. That restricts the range of things that are practical to allow in a deriving clause. In particular that rules out arbitrary expressions. I'm hoping that constraining it to constructors, literals, and variables with top-level signatures will be sufficient.

I also added a mention of `DeriveAnyClass` in that section.

tysonzero commented:
Hmm yeah that seems about right. I'm not in love with the way deriving will work, but I think the regular instance declaration is very nice.
